grafast/grafserv/src/servers/lambda/index.ts

```diff
@@ -0,0 +1,132 @@
+import type {
+  APIGatewayEvent,
+  APIGatewayProxyEvent,
+  APIGatewayProxyResult,
+  Context as LambdaContext,
+} from "aws-lambda";
+import type {
+  GrafservConfig,
+  RequestDigest} from "grafserv";
+import {
+  GrafservBase,
+  processHeaders
+} from "grafserv";
+import { toArray } from "modern-async";
+
+export class LambdaGrafserv extends GrafservBase {
+  createHandler() {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const serv = this;
+    return async function (
+      event: APIGatewayEvent,
+      context: LambdaContext,
+    ): Promise<APIGatewayProxyResult> {
+      const version = event.requestContext.protocol.match(
+        "^HTTP/(?<major>[0-9]+).(?<minor>[0-9]+)$",
+      );
+
+      const request: RequestDigest = {
+        httpVersionMajor: parseInt(version?.groups?.major ?? "1"),
+        httpVersionMinor: parseInt(version?.groups?.minor ?? "0"),
+        isSecure: event.headers["X-Forwarded-Proto"] === "https",
```

Comment:
So this is an interesting question.

Can you always trust `X-Forwarded-Proto`? No. Can you always trust `X-Forwarded-Proto` in a lambda? Maybe, but I can't find a definitive proof of it. In part, this is because there are several different ways to connect HTTP(S) to a lambda (AWS API Gateway, AWS Application Load Balancer, Amazon CloudFront, to name just the ones I am aware of), and as a result this becomes a question of "how many different teams at Amazon, probably at completely different points in time, implemented this header, and what security guarantees were they willing to make".

But let me turn this around for a second and ask a different question: why does grafserv need to know? Sure, I can see an argument that someone using grafserv with a specific server implementation (like express) might be interested in this flag, but that doesn't necessarily mean that grafserv should be in the business of exposing that flag, because I am not convinced that grafserv will ever be able to make a security guarantee about this flag that works in all server implementations. So maybe the right answer here is that the http/https flag should be exposed in the server-specific section of the request digest, rather than in the server-agnostic section, and its semantics should be left up to the specific server in question.

Reply:
Grafserv needs to know because it needs to point the client at the correct URLs for assets/etc when rendering Ruru, for example (essentially it's actually asking "is this HTTP or HTTPS in the URL?"); but we might use it for other purposes in future (e.g. deciding whether or not to send cookies, or whatever). Generally servers like Express, Fastify, etc have a setting something like "trustProxy" that can be set either to "true" or to a list of trusted endpoints. I'd be very concerned merging this without _some_ guarantee. But also, we definitely need it, so maybe if we flag it up strong enough that'll be sufficient. Is there no other way on lambda to know if it's HTTPS or not?

Reply:
The issue is that there is no such thing as "HTTPS (or not) lambda". The lambda itself receives requests on a protected backplane; this is similar to any kind of load-balancing proxy situation, in which the client only interacts with the proxy, and only the proxy interacts with the real backend.

Exposing a lambda to the outside world involves wiring up some kind of an HTTP-to-lambda bridge (of which there are at least three, as I mentioned). All of the HTTP-to-lambda bridges that I am aware of will us `X-Forwarded-Proto`, but idk that any of them have it documented as having security critical semantics. And certainly I don't feel confident enough that I know every corner of the sprawling AWS behemoth to promise that the header has robust security semantics in all possible paths from HTTP to lambda.

I think the issue here is that we are conflating two things: one is serving a web app (like Ruru) that is expecting to directly interact with the end-user-facing parts of the browser and consequently expecting to get sensible answers to browser-centric questions like "should I send cookies over" and "where are my assets" etc; the other is a microservice that serves up a node function (being the GraphQL query executor) over XHR, and doesn't need to deal with user-centric concerns at all. 

So maybe the secure flag should only be applicable to GraphiQL request handling, not to GraphQL request handling?

Reply:
For what it's worth, `serverless-express` enumerates 13 (!!) different ways to get HTTP into a lambda (https://github.com/vendia/serverless-express/tree/75e9de327972ad372a89cd7ae586b53c73c048f0/src/event-sources/aws) and it doesn't set the `req.secure` flag in any of them. 

(And yes, there's a non-zero chance that this code doesn't handle at least one of those 13 ways of getting HTTP into a lambda, and I am fine with that at this stage.)

Reply:
If we're not sure it's true, let's either:

1. set it false,
2. add a config option so the user can force it true or false, or
3. add a callback so the user can look at the incoming event and determine if it's secure or not.

Option 3 seems best; in fact it seems like the entire `const request: RequestDigest = ...` block should be provided by the _default_ implementation of this new function (as-is) and that function should be exported so that users can call it and then modify and return the result from their own callbacks. This would also enable them to do any custom tweaks for whichever of these 13 different ways they are using 😉